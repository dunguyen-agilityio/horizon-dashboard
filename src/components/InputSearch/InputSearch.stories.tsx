import { fn } from '@storybook/test';
import type { Meta, StoryObj } from '@storybook/react';

import InputSearch from '.';

const meta = {
  title: 'Components/InputSearch',
  component: InputSearch,
  tags: ['autodocs'],
  decorators: (Story) => (
    <div className="bg-white dark:bg-indigo p-4 rounded-[30px]">
      <Story />
    </div>
  ),
} satisfies Meta<typeof InputSearch>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    onSearch: fn(),
  },
};

export const WithValue: Story = {
  args: {
    onSearch: fn(),
    defaultValue: 'Sea',
  },
};
