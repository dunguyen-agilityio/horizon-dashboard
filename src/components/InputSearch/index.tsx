import React, { forwardRef } from 'react';

import SearchIcon from '@/icons/Search';
import { useInput } from '@nextui-org/input';
import { DOMElement } from '@nextui-org/system';

interface InputSearchProps {
  onSearch: (value: string) => void;
  className?: string;
  value?: string;
  defaultValue?: string;
}

const InputSearch = forwardRef<HTMLInputElement, InputSearchProps>(
  ({ onSearch, ...props }, ref) => {
    const {
      Component,
      domRef,
      startContent,
      getBaseProps,
      getInputProps,
      getInnerWrapperProps,
      getInputWrapperProps,
    } = useInput({
      ...props,
      ref,
      type: 'search',
      placeholder: 'Search...',
      startContent: (
        <span>
          <SearchIcon />
        </span>
      ),
      // custom styles
      classNames: {
        input: [
          'placeholder-purple-50 dark:placeholder-white text-primary dark:text-white',
        ],
        inputWrapper: [
          'bg-gray group-data-[focus=true]:bg-gray dark:bg-indigo-dark dark:group-data-[focus=true]:bg-indigo-dark rounded-[30px]',
        ],
        innerWrapper: ['flex gap-[11px] pl-[20px]'],
      },
    });

    const innerWrapper = React.useMemo(() => {
      const { onChange, ...rest } = getInputProps();

      const handleChange = (e: React.FormEvent<DOMElement>) => {
        onChange?.(e);
        const { value } = e.currentTarget as HTMLInputElement;
        onSearch(value);
      };

      return (
        <div {...getInnerWrapperProps()}>
          {startContent}
          <input {...rest} onChange={handleChange} data-testid="input-search" />
        </div>
      );
    }, [startContent, getInputProps, getInnerWrapperProps, onSearch]);

    const handleClickStartContent = () => {
      domRef.current?.focus();
    };

    return (
      <Component {...getBaseProps()}>
        <div
          {...getInputWrapperProps()}
          role="button"
          onClick={handleClickStartContent}
        >
          {innerWrapper}
        </div>
      </Component>
    );
  },
);

InputSearch.displayName = 'InputSearch';

export default InputSearch;
