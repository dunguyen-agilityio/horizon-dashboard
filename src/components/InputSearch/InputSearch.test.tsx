import { act, fireEvent, render, waitFor } from '@testing-library/react';
import InputSearch from '.';

describe('Input search tests', () => {
  let mockSearch: jest.Mock;

  beforeEach(() => {
    mockSearch = jest.fn();
  });

  it('Should match snapshot', () => {
    const { container } = render(<InputSearch onSearch={mockSearch} />);
    expect(container).toMatchSnapshot();
  });

  it('Should auto focus input when user click start content', async () => {
    const { getByRole, getByTestId } = render(
      <InputSearch onSearch={mockSearch} data-testid="input-search" />,
    );
    const startContent = getByRole('button');
    expect(startContent).toBeInTheDocument();
    fireEvent.click(startContent);

    await waitFor(() => {
      expect(getByTestId('input-search')).toHaveFocus();
    });
  });

  it('mockOnSearch will triggered when user typing input', async () => {
    const { getByTestId } = render(
      <InputSearch onSearch={mockSearch} data-testid="input-search" />,
    );
    const mockChange = 'value';

    act(() => {
      fireEvent.change(getByTestId('input-search'), {
        target: { value: mockChange },
      });
    });

    await waitFor(() => {
      expect(mockSearch).toHaveBeenNthCalledWith(1, mockChange);
    });
  });
});
