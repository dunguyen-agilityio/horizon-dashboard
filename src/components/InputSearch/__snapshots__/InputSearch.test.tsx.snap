// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input search tests Should match snapshot 1`] = `
<div>
  <div
    class="group flex flex-col data-[hidden=true]:hidden w-full group relative justify-end data-[has-label=true]:mt-[calc(theme(fontSize.small)_+_10px)]"
    data-filled="true"
    data-filled-within="true"
    data-slot="base"
  >
    <div
      class="relative w-full inline-flex tap-highlight-transparent flex-row items-center shadow-sm px-3 gap-3 data-[hover=true]:bg-default-200 h-10 min-h-10 transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background md:w-96 xl:w-fit bg-gray group-data-[focus=true]:bg-gray dark:bg-indigo-dark dark:group-data-[focus=true]:bg-indigo-dark rounded-[30px]"
      data-slot="input-wrapper"
      data-testid="search-wrapper"
      style="cursor: text;"
    >
      <div
        class="w-full items-center h-full box-border flex gap-[11px] pl-[8px]"
        data-slot="inner-wrapper"
      >
        <span>
          <svg
            class="[&>line]:stroke-primary [&>line]:dark:stroke-white [&>circle]:stroke-primary [&>circle]:dark:stroke-white"
            fill="none"
            height="12"
            viewBox="0 0 11 12"
            width="11"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="5"
              cy="5"
              r="4.3"
              stroke="#2B3674"
              stroke-width="1.4"
            />
            <line
              stroke="#2B3674"
              stroke-linecap="round"
              stroke-width="1.4"
              x1="10.0101"
              x2="8"
              y1="11"
              y2="8.98995"
            />
          </svg>
        </span>
        <input
          aria-label="Search"
          class="w-full font-normal bg-transparent !outline-none placeholder:text-foreground-500 focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text text-small group-data-[has-value=true]:text-default-foreground placeholder-purple-50 dark:placeholder-white text-primary dark:text-white"
          data-has-start-content="true"
          data-slot="input"
          data-testid="input-search"
          id="react-aria-:r0:"
          placeholder="Search"
          type="search"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
